### DISCUSSION TITLE
Bundling with "libc++_shared.so"

### DISCUSSION THREAD

**Discussion by kwek20** [bbqsrc/cargo-ndk#32] [Q&A]
Hi!

Ive been trying to use cargo ndk with a bigger rust library.
The library is generating Java JNI bindings through a generator called "flapigen". (So its not a direct cmake project)
After building the APK with `gradlew aR`, which triggers `gradlew jar` (For making the Java side of the generated bindings) and a gradle task for running `cargo ndk -t ${arch} -o ${libsPath} build --release`.

Yea.. a bit of a trip :D But in the end the APK gets generated, and properly links with the generated jar.
When i install the APK, it also loads the native library correctly, however, one of the native parts needs `libc++_shared.so`  (for i think a rust dependency, im in a bit over my head!).

According to .. the internet it should compile with that `.so` when i set `APP_STL:=c++_shared` in `Application.mk`, or 
```
android {
    defaultConfig {
        ...
        externalNativeBuild {
            ndkBuild {
                arguments  "-DAPP_STL=c++_shared", "-DANDROID_TOOLCHAIN=clang"
            } 
        }
    }
}
```

However, inspecting the APK, it doesnt have it incuded.

The error: `Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: library "libc++_shared.so" not found: needed by /data/app/~~1wFPIY2YxMCnGcnEq7pBpA==/org.example-MjRM-ZMHEp1x9fjGfZEjrA==/base.apk!/lib/arm64-v8a/libexample_java.so in namespace classloader-namespace`.
Ive tried building on NDK 20, 21 and 23, and targeting SDK 27, 28 and 29 to no avail.
Using gradle-6.7.1, if it makes a difference
```
        classpath 'com.android.tools.build:gradle:4.2.0'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.31"
```

Any pointers you can give me?
Thank you!

**Comment by MarijnS95** [comment]
For the record this is something that `cargo-apk` already deals with:

https://github.com/MarijnS95/android-ndk-rs/blob/670376712a2581af0c9c208e6a0ed39d7ff4fc7a/ndk-build/src/readelf.rs#L42-L45

Unfortunately `cargo-apk` cannot easily compile native binaries without the surrounding APK (see https://github.com/rust-windowing/android-ndk-rs/pull/147), which is why I envision `cargo-ndk` using `ndk-build` (`cargo-apk`'s core) as backend: https://github.com/bbqsrc/cargo-ndk/issues/28.

**Comment by rahimklaber** [comment]
Hey, did you guys figure this out? I was initially initially using https://github.com/rust-mobile/xbuild and extracting the libs from the resulting apk/aab manually. However xbuild does not support x86.

On second thought, maybe i don't need x86 support as it seems to be extremely rare.

**Reply by MarijnS95** [reply]
@rahimklaber `xbuild` and `cargo-apk`/`ndk-build` pull these directly from the NDK, so you don't even need to go through the hassle of running those tools and extracting the libs manually:

```console
a@b:~/Android/Sdk/ndk/25.1.8937393 $ fd _shared
toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_shared.so
toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/libc++_shared.so
toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/libc++_shared.so
toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/libc++_shared.so
```

That said I was under the impression `xbuild` supports at least `x86` or more specifically `x86_64`, strangely `x86`/`i686` (as well as 32-bit `arm` support) is commented out...

**Reply by rahimklaber** [reply]
Thanks!
